torch/csrc/api/src/nn/modules/conv.cpp

```diff
@@ -20,119 +25,163 @@ ConvImpl<D, Derived>::ConvImpl(const ConvOptions<D>& options_)
-    bias = this->register_parameter(
-        "bias", torch::empty(options.output_channels()));
+
+  if (options.bias()) {
+    bias = this->register_parameter("bias", torch::empty({*options.out_channels()}));
+  } else {
+    this->register_parameter("bias", Tensor(), /*requires_grad=*/false);
   }
 
-  const auto number_of_features = std::accumulate(
-      options.kernel_size()->begin(),
-      options.kernel_size()->end(),
-      options.input_channels(),
-      std::multiplies<int64_t>{});
-  const auto stdv = 1.0 / std::sqrt(number_of_features);
-  NoGradGuard no_grad;
-  for (auto& p : this->parameters()) {
-    p.uniform_(-stdv, stdv);
+  init::kaiming_uniform_(weight, /*a=*/std::sqrt(5));
+
+  if (bias.defined()) {
+    int64_t fan_in, fan_out;
+    std::tie(fan_in, fan_out) = init::_calculate_fan_in_and_fan_out(weight);
+    auto bound = 1 / std::sqrt(fan_in);
+    init::uniform_(bias, -bound, bound);
   }
 }
 
 template <size_t D, typename Derived>
 void ConvImpl<D, Derived>::pretty_print(std::ostream& stream) const {
   stream << "torch::nn::Conv" << D << "d"
-         << "(input_channels=" << options.input_channels()
-         << ", output_channels=" << options.output_channels()
-         << ", kernel_size=" << options.kernel_size()
-         << ", stride=" << options.stride() << ")";
+         << "(" << *options.in_channels()
+         << ", " << *options.out_channels()
+         << ", kernel_size=" << *options.kernel_size()
+         << ", stride=" << options.stride();
+  if (*options.padding() != *ExpandingArray<D>(0)) {
+    stream << ", padding=" << options.padding();
+  }
+  if (*options.dilation() != *ExpandingArray<D>(1)) {
+    stream << ", dilation=" << options.dilation();
+  }
+  if (*options.output_padding() != *ExpandingArray<D>(0)) {
+    stream << ", output_padding=" << options.output_padding();
+  }
+  if (options.groups() != 1) {
+    stream << ", groups=" << options.groups();
+  }
+  if (!options.bias()) {
+    stream << ", bias=" << std::boolalpha << false;
+  }
+  if (!c10::get_if<enumtype::kZeros>(&options.padding_mode())) {
+    stream << ", padding_mode=" << enumtype::get_enum_name(options.padding_mode());
+  }
+  stream << ")";
 }
 
+Conv1dImpl::Conv1dImpl(
+    ConvOptions<1> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
+
 Tensor Conv1dImpl::forward(const Tensor& input) {
-  if (options.transposed()) {
-    return torch::conv_transpose1d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.output_padding(),
-        options.groups(),
-        options.dilation());
+  if (c10::get_if<enumtype::kCircular>(&options.padding_mode())) {
+    std::vector<int64_t> expanded_padding = {((*options.padding())[0] + 1) / 2, (*options.padding())[0] / 2};
+    return F::conv1d(
+      F::pad(input, PadOptions(expanded_padding).mode(torch::kCircular)),
+      weight, bias,
+      Conv1dOptions()
```

Comment:
Why can't we just pass existing options instead recreating it?

Reply:
Thanks for the catch! Fixed.

Reply:
Sorry just realized that a call such as `options.padding(0)` is an in-place operation (i.e. it updates `padding_` of `options` instead of returning a copy of `options` that has `padding_` changed). For this reason, we will have to recreate it here because we don't actually want to change the module's `options` in place here.